### PR TITLE
Add support for stdin and stdout

### PR DESCRIPTION
**Pull request by chrisn**
- Write status messages to stderr instead of stdout
- Added --input-format and --output-format options
- Changed file I/O code to be able to read/write stdio or files

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -250,6 +253,12 @@ audio:
 
     $ audiowaveform -i test.mp3 -o test.wav
 
+You can use the `--input-format` and `--output-format` options to read from
+standard input and write to standard output. For example, the following command
+generates a waveform data file by converting a video file using ffmpeg:
+
+    $ ffmpeg -i test.mp4 -f wav - | audiowaveform --input-format wav --output-format dat -b 8 > test.dat
```
Comment (joneslee85) [inline README.md]: :+1:

File: src/AudioFileReader.cpp [hunk 2]
```diff
@@ -64,7 +64,7 @@ void AudioFileReader::showProgress(long long done, long long total)
     if (percent != percent_) {
         percent_ = percent;
 
-        output_stream << "\rDone: " << percent << "%" << std::flush;
+        error_stream << "\rDone: " << percent << "%" << std::flush;
```
Comment (joneslee85) [inline src/AudioFileReader.cpp]: the rightful place for progress info

**Comment by joneslee85** [comment]
LGTM

**Comment by chrisn** [comment]
Still to do:

- [ ] Add test cases
- [ ] Read .dat files from stdin

**Comment by chrisn** [comment]
@joneslee85 I think this PR is ready to merge. I'd welcome feedback on this, if you're interested in doing some tests.

**Comment by joneslee85** [comment]
> @joneslee85 I think this PR is ready to merge. I'd welcome feedback on this, if you're interested in doing some tests.

Code looks good. I am going to run few tests locally then come back to you.

**Comment by ffxsam** [comment]
_Very_ interested in this PR, as I'd love the ability to stream directly from AWS S3 into audiowaveform. Awesome work!

**Comment by ffxsam** [comment]
I tried this on an mp3 file, and it didn't seem to work:

```
$ cat test.mp3 | audiowaveform --no-axis-labels -e 246 --waveform-color 0073d1 --background-color 00000000 -o file.png --input-format mp3
Input file: /dev/stdin
Failed to read file: Illegal seek
```

Works fine on WAV files though.

**Comment by raedatoui** [comment]
@ffxsam how would you got about streaming S3 files, signed URLs or boto3 or something  else?

**Comment by chrisn** [comment]
Currently investigating... The `Illegal seek` error is caused by the code that skips ID3 tags in the MP3 reader.

**Comment by ffxsam** [comment]
Thanks @chrisn!

@raedatoui: Just using the AWS JavaScript SDK.

```js
const stream = s3.getObject({ Bucket: 'my-bucket', Key: 'file' }).createReadStream();
```

**Comment by chrisn** [comment]
Version 1.4.0 is now published. Thanks all for your feeback!

**Comment by tslater** [comment]
> I tried this on an mp3 file, and it didn't seem to work:
> 
> ```
> $ cat test.mp3 | audiowaveform --no-axis-labels -e 246 --waveform-color 0073d1 --background-color 00000000 -o file.png --input-format mp3
> Input file: /dev/stdin
> Failed to read file: Illegal seek
> ```
> 
> Works fine on WAV files though. And when doing `-i test.mp3`, it also works as expected.

I'm seeing the same problem with FLAC files

**Comment by chrisn** [comment]
@tslater  Please see https://github.com/bbc/audiowaveform/issues/103.

**Pull request merged**